### ISSUE TITLE
Incorrect currentBuild result on build failure inside @AfterStep hook

### ISSUE THREAD

**Issue by ravindrakmr** [jenkinsci/templating-engine-plugin#190]
### Environment Information

- [2.249.3] Jenkins version
- [ 2.0.4] Plugin version
- [Windows 10 ] OS

### Description

I am trying to send some notification on build failure but currentBuild.result is returning null to me inside @AfterStep hook method.

Here is my hook method in step_watcher.groovy
@ AfterStep
void after(){
    println "Running after the ${hookContext.library} library's ${hookContext.step} step, currentBuild.currentResult= 
   ${currentBuild.currentResult}, currentBuild.result=${currentBuild.result}"
    //steps to send notification if build is failing
}
**Pipeline console output** -
[Pipeline] // stage
[JTE][@AfterStep - hooks/step_watcher.after]
[Pipeline] echo
Running after the git library's checkout step, currentBuild.currentResult= SUCCESS, currentBuild.result=null

Basically build is failing in checkout stage but it is still showing current build result as "SUCCESS".
![image](https://user-images.githubusercontent.com/50581150/112716347-da9bfc00-8f0b-11eb-85ac-8b6ee5e267f8.png)

**Comment by steven-terrana** [comment]
Thanks for letting us know, we'll look into it and see if we can reproduce/fix.

**Comment by mariomerco** [comment]
Also happening with @CleanUp tag

**Comment by mariomerco** [comment]
Basically, is there a way right now to detect when the build failed in @Notify, @CleanUp, or similar?

**Comment by ar2rh** [comment]
Getting the same inside @CleanUp

Environment Information
**[2.277.4] Jenkins version
[ 2.2.2] Plugin version**

```
xx:xx:xx  BUILD FAILED in 4s
.
.
[JTE][@CleanUp - common/cleanup.call]
[Pipeline] echo
currentBuild.result
[Pipeline] echo
null
[Pipeline] echo
currentBuild.currentResult
[Pipeline] echo
SUCCESS
[Pipeline] End of Pipeline
hudson.remoting.ProxyException: hudson.AbortException: script returned exit code 1
.
.
.
Finished: FAILURE
```

any thoughts?

**Comment by ar2rh** [comment]
tried few version below as far as 2.0, all of them have the same issue.
For me doing try-catch is a no go.

could this issue be looked at? this problem makes notifications useless

**Comment by mxkmp** [comment]
Yes please, this prevents me from upgrading

**Comment by mxkmp** [comment]
Is there any progress? We need to upgrade from v1.7 because of the security issues for our system and this is the point why we can't upgrade our whole notification system would be not working...

**Comment by serpa-dystematic** [comment]
I'm facing this problem as well.

**Comment by mxkmp** [comment]
Is there any update on this? @steven-terrana

**Comment by zw999** [comment]
same problem

**Comment by steven-terrana** [comment]
so the problem here is actually that the build hasn't failed just because the step threw an exception. 

**Pipeline Template**
```groovy
try{
  stepThrowsException()
} catch(any){}
```

so.. `currentBuild` isn't set to failed yet.

We could add a field to `hookContext` called `stepThrewException` though. 

would that meet the need?

**Comment by steven-terrana** [comment]
I guess that would only help for lifecycle hooks that are executing after steps. 

I can make it more generic and add an `exceptionThrown` field to `hookContext`. 

so when running after Steps it'll refer to if the step that triggered the hook just threw an exception.

and when running at the end of the pipeline, it'll be triggered if there's an unresolved exception being thrown to fail the pipeline.
